File: Sorting_Searching/Topological_sort.java
import java.io.*;  
import java.util.*;  
    
// This class represents a directed graph using adjacency list representation  
class Graph  
{  
    private int V;    
  
    // Adjacency List as ArrayList 
    private ArrayList<ArrayList<Integer>> adj;  
    
    
    Graph(int v)  
    {  
        V = v;  
        adj = new ArrayList<ArrayList<Integer>>(v);  
        for (int i=0; i<v; ++i)  
            adj.add(new ArrayList<Integer>());  
    }  
    
    //adding an edge into the graph  
    void addEdge(int v,int w) { adj.get(v).add(w); }  
    
    // A recursive function used by topologicalSort  
    void topologicalSortUtil(int v, boolean visited[],Stack<Integer> stack)  
    {  
        // Mark the current node as visited.  
        visited[v] = true;  
        Integer i;  
    
        // Recur for all the vertices adjacent  
        // to thisvertex  
        Iterator<Integer> it = adj.get(v).iterator();  
        while (it.hasNext())  
        {  
            i = it.next();  
            if (!visited[i])  
                topologicalSortUtil(i, visited, stack);  
        }  
    
        // Push current vertex to stack  

        stack.push(new Integer(v));  
    }  
    
    // The function to do Topological Sort.  
    void topologicalSort()  
    {  
        Stack<Integer> stack = new Stack<Integer>();  
    
        // Mark all the vertices as not visited  
        boolean visited[] = new boolean[V];  
        for (int i = 0; i < V; i++)  
            visited[i] = false;  
    
        // Call the recursive helper    
        // Topological Sort starting from all vertices one by one  
        for (int i = 0; i < V; i++)  
            if (visited[i] == false)  
                topologicalSortUtil(i, visited, stack);  
    
        // Print contents of stack  
        while (stack.empty()==false)  
            System.out.print(stack.pop() + " ");  
    }  
    
    public static void main(String args[])  
    {  
        // Create a graph
        Graph g = new Graph(6);  
        g.addEdge(5, 2);  
        g.addEdge(5, 0);  
        g.addEdge(4, 0);  
        g.addEdge(4, 1);  
        g.addEdge(2, 3);  
        g.addEdge(3, 1);  
    
        System.out.println( "Following is a Topological sort of the given graph");
        System.out.println(g.topologicalSort());  
    }  
}  

File: Sorting_Searching/Topological_sort.py
#Python program to implement topological sorting of a DAG 
from collections import defaultdict 
  
class Graph: 
    def __init__(self,vertices): 
        self.graph = defaultdict(list)                  #dictionary containing adjacency List 
        self.V = vertices                               #No. of vertices 
  
    #  add an edge to graph 
    def addEdge(self,u,v): 
        self.graph[u].append(v) 
  
     
    def topologicalSortUtil(self,v,visited,stack): 
  
        # Mark the current node as visited. 
        visited[v] = True
  
        # Recursion for all the vertices adjacent to this vertex 
        for i in self.graph[v]: 
            if visited[i] == False: 
                self.topologicalSortUtil(i,visited,stack) 
  
        # Push current vertex to stack which stores result 
        stack.insert(0,v) 
  
    # The function to do Topological Sort 
    def topologicalSort(self): 
        # Mark all the vertices as not visited 
        visited = [False]*self.V 
        stack =[] 
  
        # Call the recursive helper function to store Topological 
        # Sort starting from all vertices one by one 
        for i in range(self.V): 
            if visited[i] == False: 
                self.topologicalSortUtil(i,visited,stack) 
  
        # Print contents of the stack 
        print stack 
  
g= Graph(6) 
g.addEdge(5, 2); 
g.addEdge(5, 0); 
g.addEdge(4, 0); 
g.addEdge(4, 1); 
g.addEdge(2, 3); 
g.addEdge(3, 1); 
  
print("Following is a Topological Sort of the given graph")
g.topologicalSort() 
